### ISSUE TITLE
UnboundLocalError: local variable 'e' referenced before assignment

### ISSUE THREAD

**Issue by nohchangsuk** [qutip/qutip#1196]
I get the following error message. My code runs fine on the first try, but gives the error if I try to run it again. I can restart the kernel every time, but would like to know if there's a fix. Could someone help please?

---------------------------------------------------------------------------
UnboundLocalError                         Traceback (most recent call last)
<ipython-input-7-becee6304ad3> in <module>
     60 
     61 
---> 62     output3 = mesolve(HDGRnew, tpss.states[i], tlist_reg, c_opsn, [Xp], args=param)
     63 
     64     Xppt = output3.expect[0]*np.exp(1j*wlo*(tlist_reg))

/opt/anaconda3/envs/qutip-env/lib/python3.8/site-packages/qutip/mesolve.py in mesolve(H, rho0, tlist, c_ops, e_ops, args, options, progress_bar, _safe_mode)
    252         ss = H
    253     elif isinstance(H, (list, Qobj, QobjEvo)):
--> 254         ss = _mesolve_QobjEvo(H, c_ops, tlist, args, options)
    255     elif callable(H):
    256         ss = _mesolve_func_td(H, c_ops, rho0, tlist, args, options)

/opt/anaconda3/envs/qutip-env/lib/python3.8/site-packages/qutip/mesolve.py in _mesolve_QobjEvo(H, c_ops, tlist, args, opt)
    298 
    299     nthread = opt.openmp_threads if opt.use_openmp else 0
--> 300     L_td.compile(omp=nthread)
    301 
    302     ss = SolverSystem()

/opt/anaconda3/envs/qutip-env/lib/python3.8/site-packages/qutip/qobjevo.py in compile(self, code, matched, dense, omp)
   1553                 if self.use_cython:
   1554                     # All factor can be compiled
-> 1555                     self.coeff_get, Code, file_ = _compiled_coeffs(
   1556                                                         self.ops,
   1557                                                         self.args,

/opt/anaconda3/envs/qutip-env/lib/python3.8/site-packages/qutip/qobjevo_codegen.py in _compiled_coeffs(ops, args, dyn_args, tlist)
    121     """
    122     code = _make_code_4_cimport(ops, args, dyn_args, tlist)
--> 123     coeff_obj, filename = _import_str(code, "cqobjevo_compiled_coeff_",
    124                                       "CompiledStrCoeff", True)
    125     return coeff_obj(ops, args, tlist, dyn_args), code, filename

/opt/anaconda3/envs/qutip-env/lib/python3.8/site-packages/qutip/qobjevo_codegen.py in _import_str(code, basefilename, obj_name, cythonfile)
     80     if not import_list:
     81         raise Exception("Could not convert string to importable function, "
---> 82                         "tmpfile:" + try_file + ext) from e
     83     coeff_obj = import_list[0]
     84     return coeff_obj, try_file + ext

UnboundLocalError: local variable 'e' referenced before assignment

Version details.
QuTiP Version:      4.5.0
Numpy Version:      1.18.1
Scipy Version:      1.3.1
Cython Version:     0.29.15
Matplotlib Version: 3.1.3
Python Version:     3.8.1
Number of CPUs:     4
BLAS Info:          INTEL MKL
OPENMP Installed:   False
INTEL MKL Ext:      True
Platform Info:      Darwin (x86_64)
Installation path:  /opt/anaconda3/envs/qutip-env/lib/python3.8/site-packages/qutip

**Comment by maij** [comment]
It looks like the loop on line 64 completes on the first try, but doesn't manage to import any objects, so the import_list remains empty. Could you use pdb to access local variables?
---
P.S. You can use Markdown to format the traceback in a user-friendly way:
e.g. 
\`\`\`python-traceback
/opt/anaconda3/envs/qutip-env/lib/python3.8/site-packages/qutip/qobjevo_codegen.py in _import_str(code, basefilename, obj_name, cythonfile)
80 if not import_list:
81 raise Exception("Could not convert string to importable function, "
---> 82 "tmpfile:" + try_file + ext) from e
83 coeff_obj = import_list[0]
84 return coeff_obj, try_file + ext
\`\`\`

Renders to:
```python-traceback
/opt/anaconda3/envs/qutip-env/lib/python3.8/site-packages/qutip/qobjevo_codegen.py in _import_str(code, basefilename, obj_name, cythonfile)
80 if not import_list:
81 raise Exception("Could not convert string to importable function, "
---> 82 "tmpfile:" + try_file + ext) from e
83 coeff_obj = import_list[0]
84 return coeff_obj, try_file + ext
```

**Comment by nohchangsuk** [comment]
Thank you for the answer. I am afraid I don't know what pdb is and how to use it. However I have somehow fixed my problem by renaming output3 to output_3. I was running back to back mesolve giving outputs which I named output2 and output3. The similarity in the names seemed to have caused the error.

**Comment by Ericgig** [comment]
The variable e does not exist outside the except block even if the try failed...
So trying to give more information about the error resulted in more confusion, sorry.

I would suggest to use functions instead of string for your coefficient, the speedpup is often less than the time it take to compile the string, (which is not included in the progressbar). 

It is the first case where it would work once, but not the following times. I would be interested in the real error message and/or a code to reproduce the bug. If you are comfortable hacking your version of qutip: adding `print(e)` in the except block
 https://github.com/qutip/qutip/blob/01132789581821517986fad1a14ab3feec7d2de2/qutip/qobjevo_codegen.py#L77-L80
should do it give the proper error message.

**Comment by nohchangsuk** [comment]
Thank you for the answer. I will try to use functions as you suggested. I didn't know how to use pdb so I didn't know what to do.

I added print(e) but nothing happens. I can attach my code (in Jupyter notebook) if you'd like to look at it.

To give you more information, I tried the same code in a Linux machine and it works fine. So I think the problem might be for Mac computers only. Also, these problem popped up after I upgraded to qutip 4.5. The problem didn't go away when I reverted to the previous Conda environment though...

**Comment by jtwamley** [comment]
Hi All
I get the same error when attempting to run the qutip tutorial 
"Bloch sphere with colorbar"

from the qutip tutorials with a virgin install of qutip from the conda-forge on Python 3.7.6 using the latest version of Anaconda..in a new environment.

Whats up?

**Comment by nathanshammah** [comment]
@jtwamley we will look into it in the next months.

**Comment by jakelishman** [comment]
Should be fixed by #1199 and released in QuTiP 4.5.2.
